include/swift/SILOptimizer/Differentiation/JVPCloner.h

```diff
@@ -29,7 +29,27 @@ namespace autodiff {
 
 class ADContext;
 
-/// A helper class for generating JVP functions.
+/// Consider a function f with type (Ti) -> U, where Ti = {T0, ..., Ti, ... Tn}
+/// and U all conform to the Differentiable protocol.
+///
+/// In that case, JVP(f) or the Jacobian Vector Product of f is defined as:
+///
+///  JVP(f)(t1) : Ti -> ((Ti) -> (U, U.TangentVector))
+///  JVP(f)(t1)(t2) = (f(Ti), Df(Ti)(t2))
+///         ^   ^--\   ^      ^--------v
+///  original args |   normal result   full derivative (jacobian) of f wrt args
+///                |
+///            free parameter passed applied to full derivative
+///
+/// In words, the JVP(f)(t1) is a function that maps the vector Ti to a 2nd
+/// function that maps t2 in T to (f(t1), Df(t1)(t2)). The first element of this
+/// result tuple, is just the constant original result of f at the point t1, but
+/// the 2nd result is a linear approximation of f at t1 that varies linearly
+/// over the input value t2.
+///
+/// Operationally, this cloner takes in a function f and produces said tuple
+/// function, producing the parts of the jacobian needed to evaluate the full
+/// derivative.
```

Comment:
Personal preference: I would use simpler writing from the ["JVP and VJP functions" section](http://bit.ly/swift-autodiff-internals) of the Swift Differentiable Programming Implementation Overview, for readers less familiar with math:
* JVP/VJP are copies of the original function that also return a differential (linear approximation function) or a pullback (backpropagation function).
* Differential takes derivative values of specific input values (corresponding to differentiation parameter indices) and returns derivative values of specific output values (corresponding to differentiation result indices).
* Pullback takes gradient values of select output values and returns gradient values of specific input values.

I find this explanation to be a bit more math-lite and explanatory of the semantics ("derivative value of input" vs "input vector").

Reply:
We could also explain the transform operationally with a simple example. [From "Derivative code generation" section](https://docs.google.com/document/d/1_BirmTqdotglwNTOcYAW-ib6mx_jl-gH9Dbg4WmHZh0/edit#heading=h.7h3piv26ukj0):

```swift
// Original function.
@differentiable
func foo(_ x: Float) -> Float {
    return sin(x) * cos(x)
}

// Simplified SIL pseudocode.
sil @foo : $(Float) -> Float {
bb0(%x):
  %y1 = apply @sin(%x)
  %y2 = apply @cos(%x)
  %y3 = apply @mul(%y1, %y2)
  return %y3
}
```

```swift
// JVP: replaces all function applications with JVP applications.
sil @jvp_foo : $(Float) -> (Float, (Float) -> Float) {
bb0(%x):
  (%y1, %df_sin) = apply @jvp_sin(%x)
  (%y2, %df_cos) = apply @jvp_cos(%x)
  (%y3, %df_mul) = apply @jvp_mul(%y1, %y2)
  // Return tuple of original result and differential.
  return (%y3, { %dx in
    %dy1 = apply %df_sin(%dx)
    %dy2 = apply %df_cos(%dx)
    %dy3 = apply %df_mul(%dy1, %dy2)
    return %dy3
  })
}
```

```swift
// VJP: replaces all function applications with VJP applications.
sil @vjp_foo : $(Float) -> (Float, (Float) -> Float) {
bb0(%x):
  (%y1, %pb_sin) = apply @vjp_sin(%x)
  (%y2, %pb_cos) = apply @vjp_cos(%x)
  (%y3, %pb_mul) = apply @vjp_mul(%y1, %y2)
  // Return tuple of original result and pullback.
  return (%y3, { %dy3 in
    // All "adjoint values" in the pullback are zero-initialized.
    // %dx = 0, %dy1 = 0, %dy2 = 0
    (%dy1, %dy2) += %pb_mul(%dy3)
    (%dx) += %pb_cos(%dy2)
    (%dx) += %pb_sin(%dy1)
    return %dx
  })
}
```

Reply:
Any thoughts here? @rxwei @gottesmm @atrick

Reply:
I'd prefer a verbatim description combined with a SIL-specific example of the function types. In the change, I find the math notation a little confusing and disconnected from the notation used in the [proposal](https://github.com/rxwei/swift-evolution/blob/autodiff/proposals/0000-differentiable-programming.md).